src/pybind/mgr/orchestrator/module.py

```diff
@@ -39,6 +39,71 @@ def to_format(what, format):
         return yaml.safe_dump_all(what, default_flow_style=False)
 
 
+def generate_preview_tables(data):
+    warning = [x.get('warning') for x in data if x.get('warning')]
+    osd_table = preview_table_osd(data)
+    service_table = preview_table_services(data)
+    tables = f"""
+{''.join(warning)}
+
+####################
```

Comment:
`PrettyTable(title='FOOSPEC PREVIEW')` doesn't work with the fork of prettytable we're using. Hence we have to use that abomination.

There is `prettytable` and `PTable`. We use `prettytable`.

Reply:
strike that. passing `hrule=HEADER` allows to add a table title/header

Reply:
Now this looks like this:

```
WARNING! Dry-Runs are only a rough estimate of what will happen.
They are snapshots of a certain period in time and are bound to the current
inventory setup. If any condition changes, the preview will be invalid. 
Please make sure to have a minimal timeframe between planning and applying the specs.
####################
SERVICESPEC PREVIEWS
####################
+---------+-------------+---------------------------+-------------+
|SERVICE  |NAME         |ADD_TO                     |REMOVE_FROM  |
+---------+-------------+---------------------------+-------------+
|mds      |mds.foo      |cephadm-dev-2 cephadm-dev  |             |
|nfs      |nfs.nfs_foo  |cephadm-dev                |             |
+---------+-------------+---------------------------+-------------+
################
OSDSPEC PREVIEWS
################
+---------+---------------+---------------+----------+----+-----+
|SERVICE  |NAME           |HOST           |DATA      |DB  |WAL  |
+---------+---------------+---------------+----------+----+-----+
|osd      |test_affinity  |cephadm-dev    |/dev/vdb  |-   |-    |
|osd      |test_affinity  |cephadm-dev    |/dev/vdc  |-   |-    |
|osd      |test_affinity  |cephadm-dev    |/dev/vdd  |-   |-    |
|osd      |test_affinity  |cephadm-dev    |/dev/vde  |-   |-    |
|osd      |test_affinity  |cephadm-dev-2  |/dev/vdb  |-   |-    |
|osd      |test_affinity  |cephadm-dev-2  |/dev/vdc  |-   |-    |
|osd      |test_affinity  |cephadm-dev-2  |/dev/vdd  |-   |-    |
|osd      |test_affinity  |cephadm-dev-2  |/dev/vde  |-   |-    |
+---------+---------------+---------------+----------+----+-----+

```